VAC/RAW/ ETH-MLS-OFFCHAIN RFC (#166)
vacp2p/rfc-index: vac/raw/eth-mls-offchain.md

---
title: ETH-MLS-OFFCHAIN
name: Secure channel setup using decentralized MLS and Ethereum accounts
status: raw
category: Standards Track
tags:
editor: Ugur Sen [ugur@status.im](mailto:ugur@status.im)
contributors: Seemenkina [ekaterina@status.im](mailto:ekaterina@status.im)
---

## Abstract

The following document specifies Ethereum authenticated scalable
and decentralized secure group messaging application by
integrating Message Layer Security (MLS) backend.

> Decentralization refers each user is a node in P2P network as well as
> each user has voice for any changes in group and decentraliation
> is achieved by integration of the consensus mechanism.

Reviewer: ```suggestion
Decentralization refers each user is a node in P2P network and
each user has voice for any changes in group.
This is achieved by integrating a consensus mechanism.
```

## Format Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”,
“SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document
are to be interpreted as described in [2119](https://www.ietf.org/rfc/rfc2119.txt).

### Assumptions

- The nodes in the P2P network can discover the nodes or they are subscribing same channel in a gossipsub.
- We MAY have non-reliable (silent) nodes.
- We MUST have a consensus that is lightweight, scalable and finalized in a specific time.

## Roles

In de-MLS RFC, there are three roles as follows:

- `node`: Nodes are members of network without being in any secure group messaging.
- `member`: Members are special nodes in the secure group messaging who
obtains current group key of secure group messaging.
- `steward`: Stewards are special and transparent members in secure group
messaging who organizes the changes upon the voted-proposals.

## MLS Background

Since the de-MLS consists of MLS backend, the MLS services and components
often used in following specification as it is or modified.

### MLS Services

MLS is operated in two services authentication service (AS) and delivery service (DS).

> AS enables group members to authenticate the credentials presented by other group members

Reviewer: ```suggestion
Authentication service enables group members to authenticate the credentials presented by other group members
```
it makes sense refer in whole name in text

> while DS routes MLS messages among the nodes or members in the protocol in the correct

Reviewer: ```suggestion
while delivery service routes MLS messages among the nodes or members in the protocol in the correct
```

order and manage the `keyPackage` of the users where the `keyPackage` is the objects
that provide some public information about a user.

### MLS Objects

Following section presents the MLS objects and components that used in this RFC:

`Epoch`: Fixed time intervals that changes the state that is defined by members,
section 3.4 in [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/).

`MLS proposal message:` Members MUST receive the proposal message prior to the
corresponding commit message that initiates a new epoch with key changes,
in order to ensure the intended security properties, section 12.1 in [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/).
In this RFC, the add and remove proposals are used.

`Application message`: This message type used in arbitrary encrypted communication between group members.
This is restricted by [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/) as if there is pending proposal,
the application message should be cut.

> Notet that: Since the MLS is based on servers, this delay between proposal and commit messages are very small.

Reviewer: ```suggestion
Note that: Since the MLS is based on servers, this delay between proposal and commit messages are very small.
```

Author: In case the readers have a question about how MLS handles this situation, so just informing them beforehand.

`Commit message:` After members receive the proposals regarding group changes,
the committer, who may be any member of the group, as specified in  [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/),
generates the necessary key material for the next epoch, including the appropriate welcome messages
for new joiners and new entropy for removed members. In this RFC, the committers only MUST be stewards.

### de-MLS Objects

`Voting proposal message:` Voting proposal messages are different then MLS proposals as voting proposals are basically
the application message in MLS group. Therefore, the steward can collect the proposals without halting the protocol.

## Flow

General flow is as follows:

> - Each time a single `steward` initializes a group and creates and emits a group anouncement (GA) periodically.

Reviewer: ```suggestion
- A steward initializes a group just once, and then sends out Group Announcements (GA) periodically..
```

> - Meanwhile, each`node`creates and sends their`credential` includes `keyPackage`.

Reviewer: ```suggestion
- Each node that wants to become a `member` of the group MUST, in response to the GA, 
share their`credential` includes `keyPackage`
```

Reviewer: still make sense

> - Each `member`creates `voting proposals` sends them to from MLS group during epoch E.

Reviewer: ```suggestion
-  During the vote, which is initiated by `steward`, each `member` of the group creates a `voting proposal` and
sends them to from MLS group during epoch `E`.
```

- Meanwhile, the `steward` collects finalized `voting proposals` from MLS group and converts them into
`MLS proposals`  then sends them with correspondng `commit messages`

> - Evantually, with the commit messages, all members starts the next epoch E+1.

Reviewer: ```suggestion
- Eventually, with the `commit message`, all members starts the next epoch `E+1`.
```

## Creating Voting Proposal

A `member` MAY initializes the voting with the proposal payload

> which is implemented using [protocol buffers v3](https://protobuf.dev/) as follows:

Reviewer: Look here I disagree, in fact only stewards start voting, then they send voting proposal and then everyone votes. 

Are we sure we want to add to the rfc that anyone can start voting?

repeated Vote votes = 14;         // Vote list in the proposal
int32 expected_voters_count = 15; // Maximum number of distinct voters
int32 round = 16;                 // Number of Votes
int64 timestamp = 17;             // Creation time of proposal
int64 expiration_time = 18;       // Time interval that the proposal is active
bool liveness_criteria_yes = 19;  // Shows how managing the silent peers vote
}
```

```bash
message Vote {
int32 vote_id = 20;             // Unique identifier of the vote
bytes vote_owner = 21;          // Voter's public key
int64 timestamp = 22;           // Time when the vote was cast
bool vote = 23;                 // Vote bool value (true/false)
bytes parent_hash = 24;         // Hash of previous owner's Vote
bytes received_hash = 25;       // Hash of previous received Vote
bytes vote_hash = 26;           // Hash of all previously defined fields in Vote
bytes signature = 27;           // Signature of vote_hash
}
```

After the member creates the voting proposal that may include adding or removing a node or member,
the voting proposal is emitted the network via the MLS application message with a lightweight,

> epoch based voting such as [hashgraphlike consensus.](https://github.com/vacp2p/rfc-index/blob/consensus-hashgraph-like/vac/raw/consensus-hashgraphlike.md)

Reviewer: not really correct 

in fact we shared message using group app topic and some separate service handle this concrete voting/proposal messages. So we can't say that we share message by voting - we share by p2p network and them use consesus service to get a result or vote

as in section 12.4.3.1. [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/).

## Single steward

To naive way to create a decentralized secure group messaging is having a single transparent `steward`
who only applies the changes regarding the result of the voting.

This is mostly similar with the general flow and specified in voting proposal and welcome message creation sections.

1. Each time a single `steward` initializes a group with group parameters with parameters
as in section 8.1. Group Context in [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/).
2. `steward` creates a group anouncement (GA) according to the previous step and
broadcast it to the all network periodically. GA message is visible in network to all `nodes`.
3. The each `node` who wants to be a member needs to obtain this anouncement and create `credential`
includes `keyPackage` that is specified in [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/) section 10.
4. The `steward` aggregates all `KeyPackages` utilizes them to provision group additions for new members,
based on the outcome of the voting process.
5. `members` start to create `voting proposals` for adding or removing a users,

> and present them to the voting in MLS group as a application message.

Reviewer: again, steward start voting

Author: addressed [here](https://github.com/vacp2p/rfc-index/pull/166/commits/389b7838ee7d96bed8006859c14e7d77ce47780d)

> 6. Meanwhile, the `steward` collects finalized `voting proposals` with in epoch E,

Reviewer: ```suggestion
6. Meanwhile, the `steward` collects finalized `voting proposals` with in epoch `E`,
```

that have received affirmative votes from members via application messages.
Otherwise, the `steward` discards proposals that did not receive a majority of "YES" votes.
Since voting proposals are transmitted as application messages, omitting them does not affect
the protocol’s correctness or consistency.
7. The `steward` converts all approved `voting proposals` into

> corresponding `MLS proposals` and `commit messages`, and

Reviewer: ```suggestion
corresponding `MLS proposals` and `commit message`, and
```

transmits both in a single operation as in [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/) section 12.4,
including welcome messages for the new members. Therefore, the `commit message` ends the previous epoch and create new ones.

## Multi stewards

Decentralization has already been achieved in the previous section.
However, to improve availability and ensure censorship resistance,
the single-steward protocol is extended to a multi-steward architecture.
In this design, each epoch is coordinated by a designated steward,
operating under the same protocol as the single-steward model.
Thus, the multi-steward approach primarily defines how steward roles
rotate across epochs while preserving the underlying structure and logic of the original protocol.
Two variants of the multi-steward design are introduced to address different system requirements.

### Multi steward with single consensus

In this model, all group modifications, such as adding or removing members,
must be approved through consensus by all participants,

> including the steward assigned for epoch E.

Reviewer: ```suggestion
including the steward assigned for epoch `E`.
```

data collection, and exclusion from decision-making processes.
To address this, we propose a decentralized, scalable peer-to-peer
group messaging system where each participant runs a node, contributes
to message propagation, and takes part in governance autonomously.
Group membership changes are decided collectively through a lightweight
partially synchronous, fault-tolerant consensus protocol without a centralized identity.
This design enables truly democratic group communication and is well-suited
for use cases like activist collectives, research collaborations, DAOs, support groups,
and decentralized social platforms.

## Format Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”,
“SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document
are to be interpreted as described in [2119](https://www.ietf.org/rfc/rfc2119.txt).

### Assumptions

- The nodes in the P2P network can discover the nodes or they are subscribing same channel in a gossipsub.
- We MAY have non-reliable (silent) nodes.
- We MUST have a consensus that is lightweight, scalable and finalized in a specific time.

## Roles

> In de-MLS RFC, there are three roles as follows:

Author: de-MLS explanation is adressed [here](https://github.com/vacp2p/rfc-index/pull/166/commits/ebb23f0627441e1ed80b695e784cb230d576d6be). 

Also we have no other specs so I updated it as in commit above.

## Single steward

To naive way to create a decentralized secure group messaging is having a single transparent `steward`
who only applies the changes regarding the result of the voting.

This is mostly similar with the general flow and specified in voting proposal and welcome message creation sections.

1. Each time a single `steward` initializes a group with group parameters with parameters
as in section 8.1. Group Context in [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/).
2. `steward` creates a group anouncement (GA) according to the previous step and
broadcast it to the all network periodically. GA message is visible in network to all `nodes`.
3. The each `node` who wants to be a member needs to obtain this anouncement and create `credential`
includes `keyPackage` that is specified in [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/) section 10.
4. The `steward` aggregates all `KeyPackages` utilizes them to provision group additions for new members,
based on the outcome of the voting process.
5. Any `member` start to create `voting proposals` for adding or removing users,
and present them to the voting in the MLS group as an application message.

> 
> However, unlimited use of `voting proposals` within the group may be misused by

Reviewer: ```suggestion
However, unlimited use of `voting proposals` within the group may be misused by
```

In this RFC, the add and remove proposals are used.

`Application message`: This message type used in arbitrary encrypted communication between group members.
This is restricted by [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/) as if there is pending proposal,
the application message should be cut.
Note that: Since the MLS is based on servers, this delay between proposal and commit messages are very small.
`Commit message:` After members receive the proposals regarding group changes,
the committer, who may be any member of the group, as specified in  [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/),
generates the necessary key material for the next epoch, including the appropriate welcome messages
for new joiners and new entropy for removed members. In this RFC, the committers only MUST be stewards.

### de-MLS Objects

`Voting proposal message:` Voting proposal messages are different then MLS proposals as voting proposals are basically
the application message in MLS group. Therefore, the steward can collect the proposals without halting the protocol.

> 

Author: Yes you are right, I think [this commit ](https://github.com/vacp2p/rfc-index/pull/166/commits/8bf3e74b7e984bfc8d2c95382e50a11cad9bdf1c) this misunderstanding

int32 proposal_id = 12;           // Unique identifier of the proposal
bytes proposal_owner = 13;        // Public key of the creator
repeated Vote votes = 14;         // Vote list in the proposal
int32 expected_voters_count = 15; // Maximum number of distinct voters
int32 round = 16;                 // Number of Votes
int64 timestamp = 17;             // Creation time of proposal
int64 expiration_time = 18;       // Time interval that the proposal is active
bool liveness_criteria_yes = 19;  // Shows how managing the silent peers vote
}
```

```bash
message Vote {
int32 vote_id = 20;             // Unique identifier of the vote
bytes vote_owner = 21;          // Voter's public key
int64 timestamp = 22;           // Time when the vote was cast
bool vote = 23;                 // Vote bool value (true/false)
bytes parent_hash = 24;         // Hash of previous owner's Vote
bytes received_hash = 25;       // Hash of previous received Vote
bytes vote_hash = 26;           // Hash of all previously defined fields in Vote
bytes signature = 27;           // Signature of vote_hash
}
```

After the member creates the voting proposal that may include adding or removing a node or member,

> the voting proposal is emitted to the network via the MLS `Application message` with a lightweight,

Author: yes fully make sense thank you for this and it is addressed [here](https://github.com/vacp2p/rfc-index/pull/166/commits/0882dcc67638fe858aeb2b36687d414d3700b7f2)

1. Each time a single `steward` initializes a group with group parameters with parameters
as in section 8.1. Group Context in [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/).
2. `steward` creates a group anouncement (GA) according to the previous step and
broadcast it to the all network periodically. GA message is visible in network to all `nodes`.
3. The each `node` who wants to be a member needs to obtain this anouncement and create `credential`
includes `keyPackage` that is specified in [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/) section 10.
4. The `steward` aggregates all `KeyPackages` utilizes them to provision group additions for new members,
based on the outcome of the voting process.
5. Any `member` start to create `voting proposals` for adding or removing users,
and present them to the voting in the MLS group as an application message.

However, unlimited use of `voting proposals` within the group may be misused by
malicious or overly active members.
Therefore, an application-level constraint can be introduced to limit the number
or frequency of proposals initiated by each member to prevent spam or abuse.

> 6. Meanwhile, the `steward` collects finalized `voting proposals` with in epoch `E`,

Author: adressed [here](https://github.com/vacp2p/rfc-index/pull/166/commits/8bf3e74b7e984bfc8d2c95382e50a11cad9bdf1c)
